src/Engine/GameEvents/TeleportEvent.java

```diff
@@ -0,0 +1,165 @@
+package Engine.GameEvents;
+
+import Engine.XYCoord;
+import Terrain.GameMap;
+import Terrain.MapMaster;
+import UI.MapView;
+import UI.Art.Animation.GameAnimation;
+import Units.Unit;
+
+/**
+ * Moves a unit directly to the destination without traversing intermediate steps.
+ * No validation is performed on the final destination, but affordances are provided for
+ * conflict resolution (e.g. existing units can swap places, die, or simply be removed).
+ * If a unit is teleported into non-traversable terrain, it will die; teleport carefully.
+ */
+public class TeleportEvent implements GameEvent
+{
+  private Unit unit;
+  XYCoord unitStart;
+  private XYCoord unitDestination;
+  private Unit obstacle;
+  private AnimationStyle animationStyle;
+  private CollisionOutcome collisionOutcome;
+
+  private boolean unitDies;
+  private boolean obstacleDies;
+
+  private GameEventQueue subEvents = new GameEventQueue();
+
+  public enum AnimationStyle
+  {
+    BLINK,
+    DROP_IN
+  }
+
+  public enum CollisionOutcome
+  {
+    KILL,
+    SWAP
+  }
+
+  public TeleportEvent(GameMap map, Unit u, XYCoord dest, AnimationStyle animStyle, CollisionOutcome crashResult)
+  {
+    unit = u;
+    unitStart = new XYCoord(unit.x, unit.y);
```

Comment:
I might default to the CO's HQ location in the case that the unit isn't on the map, for more consistent facing.

Reply:
That would make the facing wrong whenever it teleports to the wrong side of the HQ.
Maybe the "right" approach is to find the unit's Commander->HQ, figure out which side of the map it's on, and set the start (for purposes of facing, not animation) to be off the map in that direction.

Reply:
That does appear to be what you've done in Tech's case.
I don't mind that approach, I suppose.
Since the HQ facing calculation is `static` on `master`, I would like to see it reuse that function though :eyes: